### DISCUSSION TITLE
Support multiple server frameworks (instead of just express)

### DISCUSSION THREAD

**Discussion by v1rtl** [loreanvictor/tyfon#19] [Ideas]
Would be cool to have an "examples" folder with different custom servers, like Express, Polka, Koa, etc. so it'll be very fast to set up a Tyfon app with your favorite web framework, not only Express

**Comment by loreanvictor** [comment]
Good point, will make one soon (alternatively, I would appreciate a PR adding that folder ;-) ).

However, I should note that [tyfon-server](https://github.com/loreanvictor/tyfon-server/blob/main/src/run.ts) (which is the package that enables custom servers right now) only supports Express at the moment. It should be relatively easy to add support for likes of Koa and Polka (specifically at this stage), but I have the following concerns / questions about it:

1. Should support be added all in one package (e.g. tyfon-server) or should there be separate packages for each server framework? The former option would ensure no fragmentation in support for various frameworks, but the later would allow for a more iterative and agile approach (considering the fact that one of the server frameworks, currently Express, will be utilized by the CLI).

2. Adding support for more frameworks will further complicate [this](https://github.com/loreanvictor/tyfon/issues/2). With one singular server framework, you could for example just provide access to the request context for custom type / preprocessor providers, which enables using framework-specific utilities. When supporting multiple frameworks though, you can only provide a shared API for accessing request context (for example, the Node's default API which I believe all these frameworks share), which would mean you would loose access to framework-specific utilities. I am not sure how much of a problem this really is though as I would need to look into the differences of the request context in each framework.

**Comment by v1rtl** [comment]
1. Universal packages for all web frameworks at once would be cooler, so it's more flexible

2. The main reason I'd like to switch from Express because it's an ancient web framework and it doesn't provide types ootb, meaning that it's one more useless dependency. And it's slower than modern alternatives. So better would be having an option to replace express with other technologies, if needed

**Comment by loreanvictor** [comment]
> * Universal packages for all web frameworks at once would be cooler, so it's more flexible

I cannot understand the increase in flexibility:

- If all are in one package, this would be the developer workflow:
```bash
npm i tyfon-server
```
```ts
import { router } from 'tyfon-server/koa';

...
```

- If there are separate packages per web framework, this would be the developer workflow:
```ts
npm i tyfon-server-koa
```
```ts
import { router } from 'tyfon-server-koa';

...
```

Am I missing something?

> * The main reason I'd like to switch from Express because it's an ancient web framework and it doesn't provide types ootb, meaning that it's one more useless dependency. And it's slower than modern alternatives. So better would be having an option to replace express with other technologies, if needed

I understand the utility of supporting multiple web frameworks and agree with your point. Could you start creating a koa-binding package (we can later integrate it into one package if the result of that discussion is to merge all packages into one), so that we can better assess what supporting multiple web frameworks means in practice / what are the hurdles / etc? I could do that on my own, but I fear since I have not ever used koa in production or have extensive experience with it, I might not utilize it to the best of its capacities. I would of course be glad to offer any help I can on that front.

**Comment by v1rtl** [comment]
@loreanvictor what I meant is having some kind of a core package that can be easily customized to generate Koa / Polka / whatever else code

if I understand correctly, `tyfon-server` is already one

but then there should be examples with different frameworks so it's clear how to use Tyfon with them

**Comment by loreanvictor** [comment]
> @loreanvictor what I meant is having some kind of a core package that can be easily customized to generate Koa / Polka / whatever else code
> 
> if I understand correctly, `tyfon-server` is already one
> 
> but then there should be examples with different frameworks so it's clear how to use Tyfon with them

ah yes that makes perfect sense.  I think for that effort as well, first creating `tyfon-server-koa` (which is like `tyfon-server` but works with `koa` instead of `express`) would allow us to easily identify shared bits of the code that can be extracted into something like `tyfon-server-core` (p.s. already shared bits between `tyfon-server` and `tyfon-client` are extracted into `tyfon-conventions`).

**Comment by loreanvictor** [comment]
ok so I am moving this issue to [discussions](https://github.com/loreanvictor/tyfon/discussions/12) for now.
